components/webgpu/canvas_context.rs

```diff
@@ -0,0 +1,771 @@
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+//! Main process implementation of [GPUCanvasContext](https://www.w3.org/TR/webgpu/#canvas-context)
+
+use std::collections::HashMap;
+use std::ptr::NonNull;
+use std::sync::{Arc, Mutex};
+
+use arrayvec::ArrayVec;
+use base::Epoch;
+use compositing_traits::{
+    CrossProcessCompositorApi, ExternalImageSource, SerializableImageData,
+    WebrenderExternalImageApi,
+};
+use euclid::default::Size2D;
+use ipc_channel::ipc::IpcSender;
+use log::warn;
+use pixels::{IpcSnapshot, Snapshot, SnapshotAlphaMode, SnapshotPixelFormat};
+use webgpu_traits::{
+    ContextConfiguration, PRESENTATION_BUFFER_COUNT, PendingTexture, WebGPUContextId, WebGPUMsg,
+};
+use webrender_api::units::DeviceIntSize;
+use webrender_api::{
+    ExternalImageData, ExternalImageId, ExternalImageType, ImageDescriptor, ImageDescriptorFlags,
+    ImageFormat, ImageKey,
+};
+use wgpu_core::device::HostMap;
+use wgpu_core::global::Global;
+use wgpu_core::id::{
+    self, BufferId, CommandBufferId, CommandEncoderId, DeviceId, QueueId, TextureId,
+};
+use wgpu_core::resource::{
+    BufferAccessError, BufferDescriptor, BufferMapOperation, CreateBufferError,
+};
+
+use crate::wgt;
+
+pub type WGPUImageMap = Arc<Mutex<HashMap<WebGPUContextId, ContextData>>>;
+
+const fn image_data(context_id: WebGPUContextId) -> ExternalImageData {
+    ExternalImageData {
+        id: ExternalImageId(context_id.0),
+        channel_index: 0,
+        image_type: ExternalImageType::Buffer,
+        normalized_uvs: false,
+    }
+}
+
+/// Allocated buffer on GPU device
+#[derive(Clone, Copy, Debug)]
+struct Buffer {
+    device_id: DeviceId,
+    queue_id: QueueId,
+    size: u64,
+}
+
+impl Buffer {
+    /// Returns true if buffer is compatible with provided configuration
+    fn has_config(&self, config: &ContextConfiguration) -> bool {
+        config.device_id == self.device_id && self.size == config.buffer_size()
+    }
+}
+
+/// Mapped GPUBuffer
+#[derive(Debug)]
+struct MappedBuffer {
+    buffer: Buffer,
+    data: NonNull<u8>,
+    len: u64,
+    image_size: Size2D<u32>,
+    image_format: ImageFormat,
+    is_opaque: bool,
+}
+
+// Mapped buffer can be shared between safely (it's read-only)
+unsafe impl Send for MappedBuffer {}
+
+impl MappedBuffer {
+    const fn slice(&'_ self) -> &'_ [u8] {
+        // Safety: Pointer is from wgpu, and we only use it here
+        unsafe { std::slice::from_raw_parts(self.data.as_ptr(), self.len as usize) }
+    }
+
+    pub fn stride(&self) -> u32 {
+        (self.image_size.width * self.image_format.bytes_per_pixel() as u32)
+            .next_multiple_of(wgt::COPY_BYTES_PER_ROW_ALIGNMENT)
+    }
+}
+
+/// State of staging buffer
+#[derive(Debug)]
+enum StagingBufferState {
+    /// Initial state, buffer has yet to be created,
+    /// only it's id is reserved
+    Unassigned,
+    /// Buffer is allocated on GPUDevice and ready to be used immediately
```

Comment:
Is `GPUDevice` a type here? 

```suggestion
    /// The buffer is allocated on the `GPUDevice` and is ready to be used.
```